### PR TITLE
feat(code-check): enhance gb50017 skill with real computation logic

### LINKED ISSUE #52: feat(skills): define code compliance skills by standard and clause
### Body / 正文

This issue defines code compliance skills that can be split by standard, chapter, or clause. The goal is to make standards such as GB50017 and future design codes incrementally implementable.

这个 issue 用于定义规范校验技能，可按规范、章节或条文拆分，目标是让 GB50017 等规范以及未来更多设计规范能够按增量方式实现。

### Checklist / 清单

- Define how standards are represented in the skill system
- Decide whether to split by standard, chapter, or clause
- Define inputs from analysis and design stages
- Define machine-readable and user-facing outputs

- 定义规范在技能体系中的表达方式
- 决定按规范、章节还是条文拆分
- 定义来自设计与计算阶段的输入
- 定义机器可读和用户可读的输出格式

### Done When / 完成标准

- At least one standard has a clear decomposition strategy
- Compliance skills have stable inputs and outputs
- The structure supports future expansion to more standards

- 至少一种规范具备清晰拆分策略
- 规范校验技能具备稳定输入输出
- 结构可支持未来扩展到更多规范

### PR DESCRIPTION
**Pull request by guyi2000**
## Summary

- Replace stub `gb50017/code_check.py` with proper implementation that computes real utilization ratios from elementData (section/material/forces)
- Support beam, column, and brace element types with appropriate GB50017-2017 checks
- Add 46 pytest unit tests covering all check types, edge cases, and backward compatibility

## Impacted areas

- `backend` (agent-skills/code-check/gb50017/)

## What changed

**`code_check.py`** — replaced stub with:
- `get_rules()`: structured rule definitions with version `v2-member-checks`
- `_resolve_element_type()`: determine beam/column/brace from context or naming heuristics
- `_ensure_utilization_overrides()`: compute utilization ratios from elementData and inject into context
- `_check_beam()`: strength (7.1.1/7.1.2/7.1.4) + stability (8.2.1/8.4.1) + deflection (10.2.1)
- `_check_column()`: strength (7.1.1) + axial stability (8.1.1/8.4.1) + slenderness (8.3.1)
- `_check_brace()`: same as column with axial-dominant checks
- Graceful fallback to deterministic mechanism when data is incomplete

**`__tests__/test_code_check.py`** — new file with 46 tests:
- Rule structure validation
- Element type resolution (context + heuristics + defaults)
- Utilization computation for all 8 check types
- Caller override priority, partial data, zero-division safety
- Backward compatibility with existing traceability contract
- Clause reference correctness

## Test plan

- [x] 46 pytest unit tests pass
- [x] `validate_code_check_traceability` regression test passes
- [x] Backend TypeScript build succeeds
- [ ] `npm run lint --prefix backend` (CI will verify)
- [ ] Full backend regression suite (`node tests/runner.mjs check backend-regression`)

### CODE REVIEW

File: backend/src/agent-skills/code-check/gb50017/code_check.py [hunk 1]
```diff
@@ -1,39 +1,324 @@
...
+    f = material.get('f')
+    fv = material.get('fv')
+    N = forces.get('N')
+    V = forces.get('V')
+    Mx = forces.get('Mx')
+
+    # Normal stress: |N| / A / f
+    if '正应力' not in per_elem and A is not None and f is not None and N is not None:
+        try:
+            per_elem['正应力'] = abs(float(N)) / float(A) / float(f)
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-skills/code-check/gb50017/code_check.py]: Fixed in dbef129. Normal stress now includes bending contribution: `sigma = (|N|/A + |Mx|/Wnx) / f` per GB50017-2017 7.1.1.

File: backend/src/agent-skills/code-check/gb50017/code_check.py [hunk 2]
```diff
@@ -1,39 +1,324 @@
...
+        except (ZeroDivisionError, ValueError, TypeError):
+            pass
+
+    # Shear stress: |V| / A / fv
+    if '剪应力' not in per_elem and A is not None and fv is not None and V is not None:
+        try:
+            per_elem['剪应力'] = abs(float(V)) / float(A) / float(fv)
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-skills/code-check/gb50017/code_check.py]: Fixed in dbef129. Shear stress now uses `V*S/(I*tw)` per GB50017-2017 7.1.2, with `V/As` fallback when S/I/tw unavailable.

File: backend/src/agent-skills/code-check/gb50017/code_check.py [hunk 3]
```diff
@@ -1,39 +1,324 @@
...
+            pass
+
+    # Equivalent stress: sqrt(sigma^2 + 3*tau^2) / f
+    if '折算应力' not in per_elem and A is not None and f is not None and N is not None and V is not None:
+        try:
+            sigma = abs(float(N)) / float(A)
+            tau = abs(float(V)) / float(A)
+            per_elem['折算应力'] = (sigma ** 2 + 3 * tau ** 2) ** 0.5 / float(f)
+        except (ZeroDivisionError, ValueError, TypeError):
+            pass
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-skills/code-check/gb50017/code_check.py]: Fixed in dbef129. Equivalent stress now includes bending: `sqrt(sigma_axial^2 + sigma_bending^2 - sigma_axial*sigma_bending + 3*tau^2)` per GB50017-2017 7.1.4. Also added beta1 coefficient support in d9681f9.

File: backend/src/agent-skills/code-check/gb50017/code_check.py [hunk 4]
```diff
@@ -1,39 +1,324 @@
...
+                checker._calc_item(elem_id, '局部稳定', context, 'GB50017-2017 8.4.1', 'b/t ≤ [b/t]', 1.0),
+            ],
+        },
+        {
+            'name': '刚度验算',
+            'items': [
+                checker._calc_item(elem_id, '长细比', context, 'GB50017-2017 8.3.1', 'λ = l₀/i ≤ [λ]', 1.0),
+            ],
+        },
+    ]
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-skills/code-check/gb50017/code_check.py:384]: Acknowledged. Beam-column interaction checks (Clauses 8.2.2-8.2.8) are out of scope for this iteration. The current implementation covers the most common member-level checks. Combined interaction checks will be added in a follow-up when the element data schema supports the additional parameters (k factor, effective length per plane, etc.).

File: backend/src/agent-skills/code-check/gb50017/code_check.py [hunk 5]
```diff
@@ -1,39 +1,324 @@
...
+            per_elem['剪应力'] = abs(float(V)) / float(A) / float(fv)
+        except (ZeroDivisionError, ValueError, TypeError):
+            pass
+
+    # Equivalent stress: sqrt(sigma^2 + 3*tau^2) / f
+    if '折算应力' not in per_elem and A is not None and f is not None and N is not None and V is not None:
+        try:
+            sigma = abs(float(N)) / float(A)
+            tau = abs(float(V)) / float(A)
+            per_elem['折算应力'] = (sigma ** 2 + 3 * tau ** 2) ** 0.5 / float(f)
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-skills/code-check/gb50017/code_check.py]: Fixed in d9681f9. Added `beta1` parameter support from elementData (defaults to 1.0). The equivalent stress is now divided by `beta1 * f` per GB50017-2017 7.1.4.

File: backend/src/agent-skills/code-check/gb50017/code_check.py [hunk 6]
```diff
@@ -1,39 +1,324 @@
...
+                per_elem['整体稳定'] = abs(float(Mx)) / (float(phi) * float(Wnx) * float(f))
+            except (ZeroDivisionError, ValueError, TypeError):
+                pass
+
+    # Axial compression stability: |N| / (phi * A * f)
+    if '轴压稳定' not in per_elem:
+        phi = elem.get('phi')
+        if phi is not None and A is not None and f is not None and N is not None:
+            try:
+                per_elem['轴压稳定'] = abs(float(N)) / (float(phi) * float(A) * float(f))
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-skills/code-check/gb50017/code_check.py]: Fixed in d9681f9. Beam overall stability now reads `phi_b` (falls back to `phi`), and column axial stability reads `phi_axial` (falls back to `phi`). This distinguishes the two coefficients while maintaining backward compatibility.

File: backend/src/agent-skills/code-check/gb50017/code_check.py [hunk 7]
```diff
@@ -1,39 +1,324 @@
...
+                checker._calc_item(elem_id, '整体稳定', context, 'GB50017-2017 8.2.1', 'M/(φb·Wₓ·f) ≤ 1.0', 1.0),
+                checker._calc_item(elem_id, '局部稳定', context, 'GB50017-2017 8.4.1', 'b/t ≤ [b/t]', 1.0),
             ],
         },
         {
             'name': '刚度验算',
             'items': [
-                checker._calc_item(elem_id, '长细比', context, 'GB50017-2017 8.3.1', 'λ = l0/i <= λ_lim', 1.0),
-                checker._calc_item(elem_id, '挠度', context, 'GB50017-2017 10.2.1', 'f <= l/250', 1.0),
+                checker._calc_item(elem_id, '挠度', context, 'GB50017-2017 10.2.1', 'f_max ≤ L/n', 1.0),
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-skills/code-check/gb50017/code_check.py:354]: Fixed in d9681f9. Added slenderness ratio check to beam stiffness group alongside deflection. Beam now emits both `长细比` (10.1.1) and `挠度` (10.2.1).

File: backend/src/agent-skills/code-check/gb50017/__tests__/test_code_check.py [hunk 8]
```diff
@@ -0,0 +1,550 @@
...
+if _SKILL_DIR not in sys.path:
+    sys.path.insert(0, _SKILL_DIR)
+
+# Parent code_check.py dir for CodeChecker
+_PARENT_DIR = str(Path(__file__).resolve().parent.parent.parent)
+if _PARENT_DIR not in sys.path:
+    sys.path.insert(0, _PARENT_DIR)
+
+import code_check as gb50017  # noqa: E402
+
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-skills/code-check/gb50017/__tests__/test_code_check.py:20]: Not applying. The gb50017 directory is inserted first in sys.path (line 18 before line 23), so `import code_check` resolves to the gb50017 module, not the parent. All 52 tests pass with this configuration. Verified: `print(code_check.__file__)` shows the gb50017/code_check.py path.

File: backend/src/agent-skills/code-check/gb50017/__tests__/test_code_check.py [hunk 9]
```diff
@@ -0,0 +1,550 @@
...
+可直接运行: python -m pytest backend/src/agent-skills/code-check/gb50017/__tests__/test_code_check.py
+或: python backend/src/agent-skills/code-check/gb50017/__tests__/test_code_check.py
+"""
+from __future__ import annotations
+
+import sys
+from pathlib import Path
+from typing import Any, Dict, List
+from unittest.mock import MagicMock
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-skills/code-check/gb50017/__tests__/test_code_check.py]: Fixed in dbef129. The `MagicMock` import was removed in the second commit.

File: backend/src/agent-skills/code-check/gb50017/code_check.py [hunk 10]
```diff
@@ -1,39 +1,324 @@
...
+
+    section = elem.get('section', {})
+    material = elem.get('material', {})
+    forces = elem.get('forces', {})
+    if not isinstance(section, dict) or not isinstance(material, dict) or not isinstance(forces, dict):
+        return
+
+    overrides = context.setdefault('utilizationByElement', {})
+    per_elem = overrides.setdefault(elem_id, {})
+
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-skills/code-check/gb50017/code_check.py]: Fixed in dbef129. The function no longer mutates context. It now returns a new dict of computed overrides (`_compute_utilization_overrides`), and the caller merges them immutably. All `isinstance` guards are in place for the read-only access to context.

File: backend/src/agent-skills/code-check/gb50017/code_check.py [hunk 11]
```diff
@@ -1,39 +1,324 @@
...
+        lambda_limit = elem.get('lambdaLimit')
+        if l0 is not None and i_min is not None and lambda_limit is not None:
+            try:
+                per_elem['长细比'] = float(l0) / float(i_min) / float(lambda_limit)
+            except (ZeroDivisionError, ValueError, TypeError):
+                pass
+
+    # Deflection: f_max / (L / n)
+    if '挠度' not in per_elem:
+        f_max = forces.get('deflection') or elem.get('deflection')
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-skills/code-check/gb50017/code_check.py]: Fixed in d9681f9. Changed `or` to explicit `is None` check so deflection=0 is handled correctly.

File: backend/src/agent-skills/code-check/gb50017/code_check.py [hunk 12]
```diff
@@ -1,39 +1,324 @@
...
+}
+
+
+def check_element(checker: Any, elem_id: str, context: Dict[str, Any]) -> Dict[str, Any]:
+    """GB50017-2017 构件校核入口。"""
+    element_type = _resolve_element_type(elem_id, context)
+    _ensure_utilization_overrides(elem_id, context)
+    builder = _BUILDERS.get(element_type, _check_beam)
+    checks = builder(checker, elem_id, context)
+    return checker._build_element_result(elem_id, element_type, checks, 'GB50017-2017')
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-skills/code-check/gb50017/code_check.py]: Fixed in d9681f9. Unknown element types are now normalized to `'beam'` in the result via `normalized_type = element_type if element_type in _BUILDERS else 'beam'`.

File: backend/src/agent-skills/code-check/gb50017/code_check.py [hunk 13]
```diff
@@ -1,39 +1,324 @@
...
+                    },
+                    {
+                        'item': '挠度',
+                        'clause': '10.2.1',
+                        'formula': 'f_max ≤ L/n',
+                        'limit': 1.0,
+                        'params': ['deflection', 'L', 'deflectionLimitN'],
+                    },
+                ],
+            },
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-skills/code-check/gb50017/code_check.py:115]: Fixed in dbef129. The slenderness clause was changed from 8.3.1 to 10.1.1 to match the Chapter 10 group header.

File: backend/src/agent-skills/code-check/gb50017/code_check.py [hunk 14]
```diff
@@ -1,39 +1,324 @@
...
+                        'limit': 1.0,
+                        'params': ['b', 't', 'btLimit'],
+                    },
+                ],
+            },
+            {
+                'name': '刚度验算',
+                'clause': 'GB50017-2017 Chapter 10',
+                'elementType': ['beam', 'column', 'brace'],
+                'checks': [
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-skills/code-check/gb50017/code_check.py:99]: Fixed in d9681f9. Beam builder now emits the slenderness ratio check, aligning with the rule metadata.

File: backend/src/agent-skills/code-check/gb50017/code_check.py [hunk 15]
```diff
@@ -1,39 +1,324 @@
...
+        except (ZeroDivisionError, ValueError, TypeError):
+            pass
+
+    # Equivalent stress: sqrt(sigma^2 + 3*tau^2) / f
+    if '折算应力' not in per_elem and A is not None and f is not None and N is not None and V is not None:
+        try:
+            sigma = abs(float(N)) / float(A)
+            tau = abs(float(V)) / float(A)
+            per_elem['折算应力'] = (sigma ** 2 + 3 * tau ** 2) ** 0.5 / float(f)
+        except (ZeroDivisionError, ValueError, TypeError):
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-skills/code-check/gb50017/code_check.py]: Fixed in dbef129. Formulas now match the actual computation: shear uses `V*S/(I*tw)`, equivalent stress uses the full GB50017-2017 7.1.4 formula with bending, and normal stress includes the bending term.

File: backend/src/agent-skills/code-check/gb50017/__tests__/test_code_check.py [hunk 16]
```diff
@@ -0,0 +1,581 @@
...
+
+# Ensure the skill module is importable
+_SKILL_DIR = str(Path(__file__).resolve().parent.parent)
+if _SKILL_DIR not in sys.path:
+    sys.path.insert(0, _SKILL_DIR)
+
+# Parent code_check.py dir for CodeChecker
+_PARENT_DIR = str(Path(__file__).resolve().parent.parent.parent)
+if _PARENT_DIR not in sys.path:
+    sys.path.insert(0, _PARENT_DIR)
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-skills/code-check/gb50017/__tests__/test_code_check.py]: Fixed in 248e803. Changed `_PARENT_DIR` insertion from `sys.path.insert(0, ...)` to `sys.path.append(...)` so the gb50017 module always resolves first.

File: backend/src/agent-skills/code-check/gb50017/code_check.py [hunk 17]
```diff
@@ -1,39 +1,391 @@
...
+        merged_ctx = {
+            **context,
+            'utilizationByElement': {
+                **context.get('utilizationByElement', {}),
+                elem_id: {
+                    **context.get('utilizationByElement', {}).get(elem_id, {}),
+                    **computed,
+                },
+            },
+        }
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-skills/code-check/gb50017/code_check.py:477]: Fixed in 248e803. Added `isinstance(..., dict)` guards for `utilizationByElement` and the per-element entry before the immutable merge.

File: backend/src/agent-skills/code-check/gb50017/code_check.py [hunk 18]
```diff
@@ -1,39 +1,391 @@
+"""GB50017-2017 钢结构设计标准 — 构件校核实现。
+
+覆盖第 7 章（强度）、第 8 章（稳定）、第 10 章（刚度）中梁、柱、支撑的
+构件级验算。连接校核（第 9 章）不在本迭代范围内。
+"""
 from __future__ import annotations
 
-from typing import Any, Dict
+from typing import Any, Dict, List, Optional
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-skills/code-check/gb50017/code_check.py]: Fixed in 248e803. Removed unused `Optional` import.

File: backend/src/agent-skills/code-check/gb50017/code_check.py [hunk 19]
```diff
@@ -1,39 +1,397 @@
...
+                ],
+            },
+            {
+                'name': '稳定验算',
+                'clause': 'GB50017-2017 Chapter 8',
+                'elementType': ['beam', 'column', 'brace'],
+                'checks': [
+                    {
+                        'item': '整体稳定',
+                        'clause': '8.2.1',
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-skills/code-check/gb50017/code_check.py:74]: Acknowledged. Not applying because `get_rules()` serves as the catalog of all available checks across the code, while `check_element()` selects the applicable subset per element type. The `elementType` field documents which types a check *can* apply to — it is not a per-check filter. This is consistent with how other code-check skills in the project work.

File: backend/src/agent-skills/code-check/gb50017/code_check.py [hunk 20]
```diff
@@ -1,39 +1,397 @@
...
+    Wnx = section.get('Wnx') or section.get('Wx')
+
+    # Normal stress: (|N|/A + |Mx|/Wnx) / f  (GB50017-2017 7.1.1)
+    if '正应力' not in per_elem and A is not None and f is not None and N is not None:
+        try:
+            sigma_axial = abs(float(N)) / float(A)
+            sigma_bending = 0.0
+            if Mx is not None and Wnx is not None and float(Wnx) != 0:
+                sigma_bending = abs(float(Mx)) / float(Wnx)
+            computed['正应力'] = (sigma_axial + sigma_bending) / float(f)
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-skills/code-check/gb50017/code_check.py:224]: Fixed in 3e69853. Replaced `'key' not in per_elem` checks with a `_has_override()` helper that validates the existing value is numeric (`isinstance(val, (int, float))`). Non-numeric placeholders like `None` or strings no longer block computation.

File: backend/src/agent-skills/code-check/gb50017/__tests__/test_code_check.py [hunk 21]
```diff
@@ -0,0 +1,581 @@
...
+# Ensure the skill module is importable
+_SKILL_DIR = str(Path(__file__).resolve().parent.parent)
+if _SKILL_DIR not in sys.path:
+    sys.path.insert(0, _SKILL_DIR)
+
+# Parent code_check.py dir for CodeChecker (append, not insert, to avoid shadowing)
+_PARENT_DIR = str(Path(__file__).resolve().parent.parent.parent)
+if _PARENT_DIR not in sys.path:
+    sys.path.append(_PARENT_DIR)
+
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-skills/code-check/gb50017/__tests__/test_code_check.py]: Fixed in 3e69853. Removed the unused `_PARENT_DIR` path manipulation from the test file.

File: backend/src/agent-skills/code-check/gb50017/code_check.py [hunk 22]
```diff
@@ -1,39 +1,402 @@
...
+        'rules': [
+            {
+                'name': '强度验算',
+                'clause': 'GB50017-2017 Chapter 7',
+                'elementType': ['beam', 'column', 'brace'],
+                'checks': [
+                    {
+                        'item': '正应力',
+                        'clause': '7.1.1',
+                        'formula': 'σ = N/Aₙ ± Mₓ/(γₓWₙₓ) ≤ f',
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-skills/code-check/gb50017/code_check.py]: Fixed in a96d464. Updated formula in `get_rules()` from `σ = N/Aₙ ± Mₓ/(γₓWₙₓ) ≤ f` to `(|N|/A + |Mₓ|/Wₙₓ) / f ≤ 1.0` to match the actual computation.

File: backend/src/agent-skills/code-check/gb50017/code_check.py [hunk 23]
```diff
@@ -1,39 +1,402 @@
...
+                        'formula': 'τ = V·S/(I·tw) ≤ f_v',
+                        'limit': 1.0,
+                        'params': ['V', 'S', 'I', 'tw', 'fv'],
+                    },
+                    {
+                        'item': '折算应力',
+                        'clause': '7.1.4',
+                        'formula': '√(σₓ² + σ_b² - σₓ·σ_b + 3τ²) ≤ β₁·f',
+                        'limit': 1.0,
+                        'params': ['sigma_axial', 'sigma_bending', 'tau', 'f'],
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-skills/code-check/gb50017/code_check.py]: Fixed in a96d464. Added `beta1` to the `params` list for the equivalent stress rule, matching the computation in `_compute_utilization_overrides()`.

File: backend/src/agent-skills/code-check/gb50017/code_check.py [hunk 24]
```diff
@@ -1,39 +1,402 @@
...
+# Element-type-specific check builders
+# ---------------------------------------------------------------------------
+
+def _check_beam(checker: Any, elem_id: str, context: Dict[str, Any]) -> List[Dict[str, Any]]:
+    return [
+        {
+            'name': '强度验算',
+            'items': [
+                checker._calc_item(elem_id, '正应力', context, 'GB50017-2017 7.1.1', 'σ = N/Aₙ ± Mₓ/(γₓWₙₓ) ≤ f', 1.0),
+                checker._calc_item(elem_id, '剪应力', context, 'GB50017-2017 7.1.2', 'τ = V·S/(I·tw) ≤ f_v', 1.0),
```
[the remark this answers is not in the document]
Comment (guyi2000) [reply backend/src/agent-skills/code-check/gb50017/code_check.py:337]: Fixed in a96d464. Updated the beam builder formula from `σ = N/Aₙ ± Mₓ/(γₓWₙₓ) ≤ f` to `(|N|/A + |Mₓ|/Wₙₓ) / f ≤ 1.0` to match the actual utilization logic.

**Review by qinsz01 — APPROVED** [review]
